MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoInfo.java

```diff
@@ -0,0 +1,129 @@
+import com.adobe.internal.xmp.XMPMeta;
+
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.FileInputStream;
+import java.io.IOException;
+
+/**
+ * Holds data about the Motion Photo file.
+ */
+public class MotionPhotoInfo {
+
+    private final int width;
+    private final int height;
+    private final long duration;
+    private final int rotation;
+
+    private final int videoOffset;
+    private final long presentationTimestampUs;
+
+    private MediaExtractor extractor;
+
+    /**
+     * Creates a MotionPhotoInfo object associated with a given file.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    private MotionPhotoInfo(MediaFormat mediaFormat, int videoOffset, long presentationTimestampUs) {
+        width = mediaFormat.getInteger(MediaFormat.KEY_WIDTH);
+        height = mediaFormat.getInteger(MediaFormat.KEY_HEIGHT);
+        duration = mediaFormat.getLong(MediaFormat.KEY_DURATION);
+        if (mediaFormat.containsKey(MediaFormat.KEY_ROTATION)) {
+            rotation = mediaFormat.getInteger(MediaFormat.KEY_ROTATION);
+        }
+        else {
+            rotation = 0;
+        }
+
+        this.videoOffset = videoOffset;
+        this.presentationTimestampUs = presentationTimestampUs;
+    }
+
+    /**
+     * Returns a new instance of MotionPhotoInfo for a specified file.
+     */
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    public static MotionPhotoInfo newInstance(String filename) throws IOException, XMPException {
+        return MotionPhotoInfo.newInstance(filename, new MediaExtractor());
+    }
+
+    /**
+     * Returns a new instance of MotionPhotoInfo with a specified file and MediaExtractor. Used for testing.
+     */
+    @VisibleForTesting
+    @RequiresApi(api = Build.VERSION_CODES.M)
+    static MotionPhotoInfo newInstance(String filename, MediaExtractor extractor) throws IOException, XMPException {
+        XMPMeta meta = getFileXmp(filename);
+        int videoOffset = meta.getPropertyInteger("http://ns.google.com/photos/1.0/camera/", "MicroVideoOffset");
+        long presentationTimestampUs = meta.getPropertyLong("http://ns.google.com/photos/1.0/camera/", "MicroVideoPresentationTimestampUs");
+
+        MediaFormat mediaFormat = getFileMediaFormat(filename, extractor, videoOffset);
+        return new MotionPhotoInfo(mediaFormat, videoOffset, presentationTimestampUs);
+    }
+
+    /**
+     * Extract the JPEG XMP metadata from the Motion Photo.
+     */
+    private static XMPMeta getFileXmp(String filename) throws IOException, XMPException {
+        return XmpParser.getXmpMetadata(filename);
+    }
+
+    /**
+     * Get the MediaFormat associated with the video track of the Motion Photo MPEG4.
+     */
+    private static MediaFormat getFileMediaFormat(String filename, MediaExtractor extractor, int videoOffset) throws IOException {
+        File f = new File(filename);
+        try (FileInputStream fileInputStream = new FileInputStream(f)) {
```

Comment:
These two lines can be simplified as:

```

try (FileInputStream fileInputStream = new FileInputStream(new File(filename))) {
...
}
```

Reply:
I need to use the length of the file a few lines later (in the extractor.setDataSource(...) line), so that's why I stored the file in a variable for reference. Is there a way to do that if I simplify the code like this?

Reply:
ah ok.. then this is fine.